client/src/scenes/GameScene.ts

```diff
@@ -0,0 +1,224 @@
+import { RoomConnection } from "../connection";
+import { MAP, BG_COLOR } from '../../../common/map';
+
+class GameScene extends Scene {
+  private connection!: RoomConnection;
+
+  // The buffer which holds state snapshots
+  private stateBuffer: InterpolationBuffer<GameState> | undefined;
+  // A map of player sprites currently connected
+  private players: Map<string, Phaser.GameObjects.Sprite> = new Map();
+  // A map of bullet sprites currently in-air
+  private bullets: Map<number, Phaser.GameObjects.Sprite> = new Map();
+  // The graphics representation of the map objects
+  private mapGfx: Phaser.GameObjects.Graphics | undefined;
+  // The Hathora user for the current client's connected player
+  private hathoraUser: object & { id: string } | undefined;
+  // The current client's connected player's sprite object
+  private playerSprite: Phaser.GameObjects.Sprite | undefined;
+  // The previous tick's aim radians (used to check if aim has changed, before sending an update)
+  private prevAimRad: number = 0;
+
+  constructor() {
+    super("scene-game");
+  }
+
+  init(data: Record<string, any>) {
+    // Receive connection and user data from BootScene
+    this.connection = data.connection;
+    this.hathoraUser = data.hathoraUser;
+  }
+
+  create() {
+    this.connection.addListener(({ state, ts }) => {
+      // Start enqueuing state updates
+      if (this.stateBuffer === undefined) {
+        this.stateBuffer = new InterpolationBuffer(state, 50, lerp);
+      } else {
+        this.stateBuffer.enqueue(state, [], ts);
+      }
+    });
+
+    // Handle keyboard input
+    const keys = this.input.keyboard.addKeys('W,S,A,D') as {
+      W: Phaser.Input.Keyboard.Key,
+      S: Phaser.Input.Keyboard.Key,
+      A: Phaser.Input.Keyboard.Key,
+      D: Phaser.Input.Keyboard.Key
+    };
+    let prevDirection = Direction.None;
+
+    const handleKeyEvt = () => {
+      let direction: Direction;
+      if (keys.W.isDown) {
+        direction = Direction.Up;
+      } else if (keys.S.isDown) {
+        direction = Direction.Down;
+      } else if (keys.D.isDown) {
+        direction = Direction.Right;
+      } else if (keys.A.isDown) {
+        direction = Direction.Left;
+      } else {
+        direction = Direction.None;
+      }
+
+      if (prevDirection !== direction) {
+        // If connection is open and direction has changed, send updated direction
+        prevDirection = direction;
+        this.connection.sendMessage({ type: ClientMessageType.SetDirection, direction });
+      }
+    };
+
+    this.input.keyboard.on("keydown", handleKeyEvt);
+    this.input.keyboard.on("keyup", handleKeyEvt);
+
+    // Handle mouse-click input
+    this.input.on(Phaser.Input.Events.POINTER_DOWN, () => {
+      // If the connection is open, send through click events
+      this.connection.sendMessage({ type: ClientMessageType.Shoot });
+    });
+
+    // Render map objects
+    this.mapGfx = this.add.graphics();
```

Comment:
this is only used here, so it can be a local variable instead of a class one

Reply:
Good catch, at one point I think I was planning on changing it in the `update()` loop (moving objects).